### PR TITLE
Explain the lack of flow IDs, and suggest a pattern

### LINKED ISSUE #6: No streams in datagrams?
I finally sat down and read this draft (nice work, BTW) and was somewhat sad to learn that QUIC won't natively support stream multiplexing for datagrams.

It is true that applications can implement this if they choose. But one nice thing about QUIC is that it takes care of this machinery on behalf of the application. This is a somewhat aesthetic concern, but it feels awkward to have one layer of stream multiplexing in QUIC and then another layer in the application. It would be ugly in our implementation, at least.

Moving on to non-aesthetic issues: I don't have a full grasp of all the use cases for DATAGRAM, but if stream multiplexing is a common requirement, it would be good to move this into the transport. Applications that don't need this are free to do it all over one stream at a tiny loss in wire efficiency, or we could have a stream-less version if people are deeply concerned about one byte per frame.

### PR DESCRIPTION
**Pull request by tfpauly**
Closes #6

### CODE REVIEW

File: draft-ietf-quic-datagram.md [hunk 1]
```diff
@@ -206,20 +206,32 @@ and can store the contents in memory.
...
+Identifiers used to multiplex different kinds of datagrams, or flows of
+datagrams, are the responsibility of the application protocol running over QUIC
+to define. The application defines the semantics of the Datagram Data field
+and how it is parsed.
+
+If the application needs to support the coexistence of multiple flows of
+datagrams, one recommended pattern is to use a variable-length integer at the
+beginning of the Datagram Data field. For example, using datagrams with
+HTTP/3 involves prepending such a flow identifier to all datagrams
+(see {{?I-D.schinazi-quic-h3-datagram}}).
```
Comment (nibanks) [inline draft-ietf-quic-datagram.md]: I assume H3 datagram spec also references back to this one? Does it make sense to have a circular reference? I know some in the IETF discussion wanted this text, but I'm personally fine with out this last paragraph.

**Review by nibanks — APPROVED** [review]

File: draft-ietf-quic-datagram.md (hunk 1 above)
Comment (DavidSchinazi) [reply draft-ietf-quic-datagram.md]: Circular references are common and fine

**Review by DavidSchinazi — APPROVED** [review]

File: draft-ietf-quic-datagram.md (hunk 1 above)
Comment (tfpauly) [reply draft-ietf-quic-datagram.md]: Yup, this is a non-normative reference; H3 datagram will have a normative reference.

Comment (DavidSchinazi) [reply draft-ietf-quic-datagram.md]: Agreed, though FWIW circular normative references are OK - they just require simultaneous publication.

Comment (LPardue) [reply draft-ietf-quic-datagram.md]: I'm not too bothered by a circular reference but I don't think it adds much. The wire format is simple. The main complexity of the H3 draft comes from how the flow IDs meld with HTTP semantics. So if you're going to reference it then be more specific about why, or just don't bother IMO.

And if you are going to reference it, let's use the MASQUE edited draft

Comment (LPardue) [reply draft-ietf-quic-datagram.md]: Given some of the churn in H3 DGRAM, how about we extract this third paragraph into another issue/PR that we can sit on while we land the other changes?

**Comment by martinduke** [comment]
I would like this section to say something about priorities, as emerged in the issue. Something to the effect of:

"QUIC Implementations SHOULD present an API to applications to assign relative priorities to DATAGRAM frames with respect to each other and to QUIC streams. This can be done by the API conceptually assigning DATAGRAM frames to an existing QUIC stream and inheriting its priority, or assigning a priority in the DATAGRAM send API itself."

**Comment by LPardue** [comment]
the first sentence of @martinduke's suggestion sounds ok. I don't like the second one much (sorry).

edit: clarification - the first sentence is very much like RFC 9000's "A QUIC implementation SHOULD provide ways in which an application can indicate the relative priority of streams"

**Comment by martinduke** [comment]
> 
> 
> the first sentence of @martinduke's suggestion sounds ok. I don't like the second one much (sorry).
> 
> edit: clarification - the first sentence is very much like RFC 9000's "A QUIC implementation SHOULD provide ways in which an application can indicate the relative priority of streams"

I'm not married to the second sentence.

**Comment by DavidSchinazi** [comment]
Piling on to agree here: I'm OK with adding the first sentence, though I think the second one is too specific so I'd prefer the second sentence wasn't added

**Comment by tfpauly** [comment]
Added the first sentence about priorities. Please check!

File: draft-ietf-quic-datagram.md (hunk 1 above)
Comment (tfpauly) [reply draft-ietf-quic-datagram.md]: @LPardue removed the specific H3 datagram reference; that OK?

File: draft-ietf-quic-datagram.md [hunk 2]
```diff
@@ -206,20 +206,33 @@ and can store the contents in memory.
...
 the max_packet_size transport parameter, and by the Maximum Transmission Unit
 (MTU) of the path between endpoints. DATAGRAM frames cannot be fragmented,
 therefore application protocols need to handle cases where the maximum datagram
 size is limited by other factors.
 
+## Multiplexing Datagrams
+
+DATAGRAM frames belong to a QUIC connection as a whole, and are not strongly
+associated with any stream ID at the QUIC layer. However, it is expected
+that applications will want to differentiate between different DATAGRAM frames
```
Comment (LPardue) [inline draft-ietf-quic-datagram.md]: "differentiate between different" seems kind of redundant. I tried penning a suggestion for this sentence but it didn't seem to read any better

**Review by LPardue — APPROVED** [review]
I made a grammatical nit, but in the absence being able to make a firm suggestion, this LGTM.

File: draft-ietf-quic-datagram.md (hunk 1 above)
Comment (LPardue) [reply draft-ietf-quic-datagram.md]: yep!

**Review by martinduke — APPROVED** [review]
LGTM

File: draft-ietf-quic-datagram.md (hunk 2 above)
Comment (tfpauly) [reply draft-ietf-quic-datagram.md]:
```suggestion
that applications will want to differentiate between specific DATAGRAM frames
```

**Pull request merged**